src/Framing/Attach.cs

```diff
@@ -32,132 +32,237 @@ public Attach()
+            get { return this.handle == null ? uint.MinValue : this.handle.Value; }
+            set { this.handle = value; }
         }
 
+        private bool? role;
         /// <summary>
         /// Gets or sets the role field.
         /// </summary>
         public bool Role
         {
-            get { return this.Fields[2] == null ? false : (bool)this.Fields[2]; }
-            set { this.Fields[2] = value; }
+            get { return this.role == null ? false : this.role.Value; }
+            set { this.role = value; }
         }
 
+        private SenderSettleMode? sndSettleMode;
         /// <summary>
         /// Gets or sets the snd-settle-mode field.
         /// </summary>
         public SenderSettleMode SndSettleMode
         {
-            get { return this.Fields[3] == null ? SenderSettleMode.Unsettled : (SenderSettleMode)this.Fields[3]; }
-            set { this.Fields[3] = (byte)value; }
+            get { return this.sndSettleMode == null ? SenderSettleMode.Unsettled : this.sndSettleMode.Value; }
+            set { this.sndSettleMode = value; }
         }
 
+        private ReceiverSettleMode? rcvSettleMode;
         /// <summary>
         /// Gets or sets the rcv-settle-mode field.
         /// </summary>
         public ReceiverSettleMode RcvSettleMode
         {
-            get { return this.Fields[4] == null ? ReceiverSettleMode.First : (ReceiverSettleMode)this.Fields[4]; }
-            set { this.Fields[4] = (byte)value; }
+            get { return this.rcvSettleMode == null ? ReceiverSettleMode.First : this.rcvSettleMode.Value; }
+            set { this.rcvSettleMode = value; }
         }
 
+        private object source;
         /// <summary>
         /// Gets or sets the source field.
         /// </summary>
         public object Source
         {
-            get { return this.Fields[5]; }
-            set { this.Fields[5] = value; }
+            get { return this.source; }
+            set { this.source = value; }
         }
 
+        private object target;
         /// <summary>
         /// Gets or sets the target field.
         /// </summary>
         public object Target
         {
-            get { return this.Fields[6]; }
-            set { this.Fields[6] = value; }
+            get { return this.target; }
+            set { this.target = value; }
         }
 
+        private Map unsettled;
         /// <summary>
         /// Gets or sets the unsettled field.
         /// </summary>
         public Map Unsettled
         {
-            get { return (Map)this.Fields[7]; }
-            set { this.Fields[7] = value; }
+            get { return this.unsettled; }
+            set { this.unsettled = value; }
         }
 
+        private bool? incompleteUnsettled;
         /// <summary>
         /// Gets or sets the incomplete-unsettled field.
         /// </summary>
         public bool IncompleteUnsettled
         {
-            get { return this.Fields[8] == null ? false : (bool)this.Fields[8]; }
```

Comment:
Why not to use auto props, as it is done in azure-amqp implementation?

Reply:
The backing field needs to be nullable but I didn't want to make the public properties nullable as that would be a breaking change. Additionally many properties have a logic for returning a default value if the backing field is null, but the field shouldn't be initialized with this default value as that would change the encoded value if the field is not used.